evalml/tests/automl_tests/test_automl_iterative_algorithm.py

```diff
@@ -425,8 +426,13 @@ def test_max_batches_num_pipelines(
             + num_ensemble_batches
         )
         n_automl_pipelines = n_results
-    assert automl._automl_algorithm.batch_number == max_batches
-    assert automl._automl_algorithm.pipeline_number + 1 == n_automl_pipelines
+    if max_batches is None:
```

Comment:
This is a weird one and let me try to explain! In this PR, I make[ these changes ](https://github.com/alteryx/evalml/pull/3304/files#diff-8b4df814812e6f5df24c600bc7f429b036e0e2f78613bfbdcdf0ed3ab86bffd6R732-R736) which not only set the default behavior for max batches but more importantly moved this logic after the automl algorithm is instantiated. So, on main, the default max batches is set before the automl algorithm is set. This meant that if `max_batches`, `max_iterations` etc. were all `None` on main, `max_batches` would be set to 1 and would be passed into `IterativeAlgorithm`. However, `IterativeAlgorithm` has [this logic ](https://github.com/alteryx/evalml/blob/main/evalml/automl/automl_algorithm/iterative_algorithm.py#L227-L254) which sets `max_iterations` if `ensembling = True` and if `max_batches is not None`. This meant that on `main`  , turning `ensembling=True` would mean that `AutoMLSearch` would use `max_iterations` as stopping criteria. When I made the changes to search in this PR, it meant that search was correctly using `max_batches` as stopping criteria which meant that search would stop when `batch_number > max_batches`. These changes to this test reflect that and with an added complication of `_pipelines_per_batch`.

Reply:
Think the above makes sense but still have some details to iron out.